pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1459 @@
+
+			Self::update_imbalance(current_imbalance, state_changes.delta_imbalance)?;
+
+			Self::update_hdx_subpool_hub_asset(state_changes.hdx_hub_amount)?;
+
+			Self::set_asset_state(asset_in, new_asset_in_state);
+			Self::set_asset_state(asset_out, new_asset_out_state);
+
+			Self::deposit_event(Event::SellExecuted {
+				who,
+				asset_in,
+				asset_out,
+				amount_in: amount,
+				amount_out: *state_changes.asset_out.delta_reserve,
+			});
+
+			Ok(())
+		}
+
+		/// Execute a swap of `asset_out` for `asset_in`.
+		///
+		/// Price is determined by the Omnipool.
+		///
+		/// Hub asset is traded separately.
+		///
+		/// Asset's tradable states must contain SELL flag for asset_in and BUY flag for asset_out, otherwise `NotAllowed` error is returned.
+		///
+		/// Parameters:
+		/// - `asset_in`: ID of asset sold to the pool
+		/// - `asset_out`: ID of asset bought from the pool
+		/// - `amount`: Amount of asset sold
+		/// - `max_sell_amount`: Maximum amount to be sold.
+		///
+		/// Emits `BuyExecuted` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::buy())]
+		#[transactional]
+		pub fn buy(
+			origin: OriginFor<T>,
+			asset_out: T::AssetId,
+			asset_in: T::AssetId,
+			amount: Balance,
+			max_sell_amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinimumTradingLimit::get(),
+				Error::<T>::InsufficientTradingAmount
+			);
+
+			// Special handling when one of the asset is Hub Asset
+			if asset_out == T::HubAssetId::get() {
+				return Self::buy_hub_asset(&who, asset_in, amount, max_sell_amount);
+			}
+
+			if asset_in == T::HubAssetId::get() {
+				return Self::buy_asset_for_hub_asset(&who, asset_out, amount, max_sell_amount);
+			}
+
+			let asset_in_state = Self::load_asset_state(asset_in)?;
+			let asset_out_state = Self::load_asset_state(asset_out)?;
+
+			ensure!(
+				Self::allow_assets(&asset_in_state, &asset_out_state),
+				Error::<T>::NotAllowed
+			);
+
+			ensure!(asset_out_state.reserve >= amount, Error::<T>::InsufficientLiquidity);
+
+			let current_imbalance = <HubAssetImbalance<T>>::get();
+
+			let state_changes = hydra_dx_math::omnipool::calculate_buy_state_changes(
+				&(&asset_in_state).into(),
+				&(&asset_out_state).into(),
+				amount,
+				T::AssetFee::get(),
+				T::ProtocolFee::get(),
+				current_imbalance.value,
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			ensure!(
+				T::Currency::free_balance(asset_in, &who) >= *state_changes.asset_in.delta_reserve,
+				Error::<T>::InsufficientBalance
+			);
+
+			ensure!(
+				*state_changes.asset_in.delta_reserve <= max_sell_amount,
+				Error::<T>::SellLimitExceeded
+			);
+
+			let new_asset_in_state = asset_in_state
+				.delta_update(&state_changes.asset_in)
+				.ok_or(ArithmeticError::Overflow)?;
+			let new_asset_out_state = asset_out_state
+				.delta_update(&state_changes.asset_out)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			T::Currency::transfer(
```

Comment:
deltas of the actual transfers should be probably checked, this can be resolved easily by wrapping the transfer function